x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -152,27 +153,57 @@ protected Query doToQuery(SearchExecutionContext context) throws IOException {
         if (ft == null) {
             return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
         }
+
+        final String fieldTypeName = ft.typeName();
+        if (AllowedFieldType.isFieldTypeAllowed(fieldTypeName) == false) {
+            throw new ElasticsearchParseException(
+                "["
+                    + fieldTypeName
+                    + "]"
+                    + " is not an appropriate field type for text expansion query. "
+                    + "Allowed field types are ["
+                    + AllowedFieldType.getAllowedFieldTypesAsString()
+                    + "]."
+            );
+        }
+
+        return (this.tokenPruningConfig == null)
+            ? queryBuilderWithAllTokens(tokens, ft, context)
+            : queryBuilderWithPrunedTokens(tokens, ft, context);
+    }
+
+    private Query queryBuilderWithAllTokens(List<WeightedToken> tokens, MappedFieldType ft, SearchExecutionContext context) {
         var qb = new BooleanQuery.Builder();
-        int fieldDocCount = context.getIndexReader().getDocCount(fieldName);
-        float bestWeight = 0f;
-        for (var t : tokens) {
-            bestWeight = Math.max(t.weight(), bestWeight);
+
+        for (var token : tokens) {
+            qb.add(new BoostQuery(ft.termQuery(token.token(), context), token.weight()), BooleanClause.Occur.SHOULD);
         }
+        return qb.setMinimumNumberShouldMatch(1).build();
+    }
+
+    private Query queryBuilderWithPrunedTokens(List<WeightedToken> tokens, MappedFieldType ft, SearchExecutionContext context)
+        throws IOException {
+        var qb = new BooleanQuery.Builder();
+        int fieldDocCount = context.getIndexReader().getDocCount(fieldName);
+        float bestWeight = findBestWeightFor(tokens);
         float averageTokenFreqRatio = getAverageTokenFreqRatio(context.getIndexReader(), fieldDocCount);
         if (averageTokenFreqRatio == 0) {
             return new MatchNoDocsQuery("The \"" + getName() + "\" query is against an empty field");
         }
+
         for (var token : tokens) {
             boolean keep = shouldKeepToken(context.getIndexReader(), token, fieldDocCount, averageTokenFreqRatio, bestWeight);
-            if (this.tokenPruningConfig != null) {
-                keep ^= this.tokenPruningConfig.isOnlyScorePrunedTokens();
-            }
+            keep ^= this.tokenPruningConfig.isOnlyScorePrunedTokens();
             if (keep) {
                 qb.add(new BoostQuery(ft.termQuery(token.token(), context), token.weight()), BooleanClause.Occur.SHOULD);
             }
         }
-        qb.setMinimumNumberShouldMatch(1);
-        return qb.build();
+
+        return qb.setMinimumNumberShouldMatch(1).build();
+    }
+
+    private float findBestWeightFor(List<WeightedToken> tokens) {
```

Comment:
Nice cleanup here on the for loop. Since it's one line though it may not need to be its own method? You could just do `float bestWeight = tokens.stream().map(WeightedToken::weight).reduce(0f, Math::max);`

Reply:
Updated